### PR TITLE
Timeout jwt

### LINKED ISSUE #2: Refresh JWT token before timeout
The JWT tokens used with GCP times out. The timeout is configurable, but only up to a maximum of 24 hrs according to [docs](https://cloud.google.com/iot/docs/how-tos/credentials/jwts#jwt_claims). My main issue is that when the timeout occurs, I get errors in my dashboard, even though the client immediately reconnects and it's not a problem for the device. Having a timer that kicks in about 30 seconds before the token expires, and refreshing it is probably a good solution to this issue. It should not cause much extra work for the devices, because when the token expires, the connection is dropped, and the client has to re-generate a token regardless.

### PR DESCRIPTION
**Pull request by Alxandr**
I'm new to a lot of this, so I'd like some early review if anyone could provide it. This is not yet tested, so it might very well segfault (I'm doing testing atm).

Closes #2

### CODE REVIEW

File: src/mgos_gcp.c [hunk 1]
```diff
@@ -180,7 +223,12 @@ bool mgos_gcp_init(void) {
     LOG(LL_ERROR, ("Invalid gcp.key (0x%x)", r));
     return false;
   }
-  mgos_mqtt_set_connect_fn(mgos_gcp_mqtt_connect, NULL);
+
+  struct gcp_state *state = malloc(sizeof(*state));
```
Comment (rojer) [inline src/mgos_gcp.c]: always use `calloc` to allocate structs

File: src/mgos_gcp.c [hunk 2]
```diff
@@ -66,10 +72,23 @@ static void base64url_putc(char c, void *arg) {
   mbuf_append(&ctx->jwt, &c, 1);
 }
 
+static void mgos_gcp_jwt_timeout(void *param) {
+  struct gcp_state *state = (struct gcp_state *)param;
+  state->jwt_timeout = 0;
```
Comment (rojer) [inline src/mgos_gcp.c]: use MGOS_INVALID_TIMER_ID instead of 0, here and in other places

File: src/mgos_gcp.c [hunk 3]
```diff
@@ -66,10 +72,23 @@ static void base64url_putc(char c, void *arg) {
...
+  LOG(LL_DEBUG, ("Disconnecting global MQTT connection due to JWT timeout."));
```
Comment (rojer) [inline src/mgos_gcp.c]: make it INFO level and change wording somewhat: "Dropping MQTT connection due to imminent token expiration"

File: src/mgos_gcp.c [hunk 4]
```diff
@@ -44,6 +46,10 @@ struct jwt_printer_ctx {
   struct mbuf jwt;
 };
 
+struct gcp_state {
+  mgos_timer_id jwt_timeout;
```
Comment (rojer) [inline src/mgos_gcp.c]: `token_ttl_timer_id`

File: src/mgos_gcp.c [hunk 5]
```diff
@@ -82,11 +101,15 @@ static void mgos_gcp_mqtt_connect(struct mg_connection *c,
...
   base64url_putc('.', &ctx);
   uint64_t iat = (uint64_t) now;
-  uint64_t exp = iat + mgos_sys_config_get_gcp_token_ttl();
+  uint64_t ttl = (uint64_t) mgos_sys_config_get_gcp_token_ttl();
+  uint64_t exp = iat + ttl;
   if (exp < 1500000000) {
     LOG(LL_ERROR, ("Time is not set, GCP connection will fail. "
                    "Set the time or make sure SNTP is enabled and working."));
   }
+  state->jwt_timeout = mgos_set_timer((ttl * 1000) - 30000, 0,
```
Comment (rojer) [inline src/mgos_gcp.c]: `(ttl - 30) * 1000`

**Review by rojer — CHANGES_REQUESTED** [review]
a few stylistic changes, otherwise LGTM.
also please run `clang-format` with [this config](https://github.com/cesanta/mongoose-os/blob/master/.clang-format). on the file.

**Comment by Alxandr** [comment]
Updated based on comments, however the timer does not seem to trigger. Any idea what could be causing this?

**Comment by Alxandr** [comment]
I figured it out. I had setup of my project wrong, so I was building using the normal gcp version, and not my fork. I've fixed it, and can see that it's working (just setting the `ttl` to 120).

**Comment by Alxandr** [comment]
Having run this for a night, I don't see any problems. So it seems the issue is solved :)

**Comment by rojer** [comment]
ok, this looks good, integrating. thank you!

note: you will see this PR closed without merging and a different commit will appear instead.

**Comment by Alxandr** [comment]
Out of curiosity; why on earth are you commiting other peoples code as your own?

**Comment by rojer** [comment]
@Alxandr this is because commits first go to our internal monolithic repo and then get published to individual lib repos by a bot. we don't merge PRs into external repos directly because then they'll get out of sync and we'll have merge conflicts during publishing, which bot cannot resolve.

**Comment by Alxandr** [comment]
So what you're saying is that you've made a system, in which contributors do not get attribution as part of the history? Doesn't that seem a bit disingenuous? Now, honestly, regarding this code here, I don't really care, and you're free to use it however you like; but I was not asked about this. There are no contributor guidelines that state that if you make a PR we will just copy the code and claim it as our own. From my perspective at least, if you want contributors, I'd strongly suggest rethinking this strategy (and/or any automation you've built that enforces it).

**Comment by rojer** [comment]
@Alxandr we always reference the original PR in the description, and usually i also explicitly credit the author by @-mention - see e.g. https://github.com/cesanta/mongoose-os/commit/c9d76ac82f1058cd4a14de84da3964b37aa4fe1e
i forgot to do it this time, my bad.

this system is what we had in place since day 1, even before we started accepting contributions and all work was done in house.
it is our plan to stop this publishing system, and we are working towards that, but it is not yet possible (there is a legacy internal CI system that depends on having everything in one repo).
please bear with us here, i agree this is sub-optimal.

**Comment by Alxandr** [comment]
Yeah, as said, in this case I don't really care, cause it was a one off fix. But if I was going to contribute any significant time to a project, it would definitely be off-putting.

**Pull request closed without merging**